### PR TITLE
server: tell a streaming client when its slot is parked and restored

### PR DESCRIPTION
**Pull request by danielhanchen**
Stacked on #184. Only the last commit is new; the rest is the swap branch.

## Summary

A streaming client is told when its slot is parked and when it is restored, as SSE comments (`: preempted`, `: resumed`), with a `: preempt-keepalive` every 2 s while parked. Comments are legal SSE and invisible to every existing client, so the body is unchanged by preemption. Non-streaming requests see nothing. A request parked before its first token starts its stream on the notice, so the headers and the keepalive go out through the wait instead of after it.

Without this a parked slot is a silent connection: the client cannot tell "waiting for cells" from "slow" or "dead", and any client-side read timeout or proxy idle timeout fires during a wait that is long by design. Unsloth Studio uses the comments to show a paused line on the chat that was parked and to stand its own preemption down when the server can do it.

## Policy

Unchanged. The notice is emitted at the two park sites (the KV-full victim and the `LLAMA_SERVER_PREEMPT_EVERY` test knob) and at the restore site, through a `server_task_result_preempt_notice` pushed to the task's result queue for streaming tasks only. Notices that precede the first result (a slot parked while still processing its prompt) are sent in front of it. While parked the ping period is 2 s regardless of `--sse-ping`.

## Results

Raw stream under `LLAMA_SERVER_PREEMPT_EVERY=64`: park and resume pairs every 64 tokens, body identical to the unparked run. Through Unsloth Studio on Qwen3.5-4B with two MTP drafts at `-c 8192` and four slots: four concurrent chats 4 of 4 with the pause shown in the GUI on every chat that was parked, zero errors.

## Exactness

The generated tokens with forced parks equal the unparked run token for token (harness test), and through Studio both 1200 token prompts are byte-identical across 18 parks each (seed 0, temperature 0).

## Cost when it does not fire

None. No notice is created unless a slot is parked, and the keepalive is the existing `--sse-ping` path with one extra branch.

## Tests

`tools/server/tests/unit/test_preempt_notify.py`: comments arrive in order and the body is unchanged on `/completion`, the same on `/v1/chat/completions`, a non-streaming request is untouched, two overflowing streams both finish with the parked one saying so, a stream parked before its first token starts with the notice and the notice arrives while the other stream still runs, and a resident rotated out for a parked head announces it with every notice paired. 17 passed together with the existing preemption tests.

## Limitations

No keepalive before the first result, so a park during prompt processing is covered by the client's first-token deadline rather than the 2 s ping.

## Follow-up commit

The keepalive promised above had never fired. The server's result queue notifies one condition variable for every result of every task, and the reader's timed wait restarted on each wakeup, so on any busy server the timeout never elapsed; the same wait serves the ordinary `--sse-ping`. The wait now targets a fixed deadline. Paired on this branch against an isolated build of the previous head, a 13.9 s park carried zero keepalives before and four after, spaced at the 2 s period plus the reader's 1 s polling granularity, none outside the park.

Carried from #184 since: a pure recurrent cache is served without preemption (no notices either, nothing is parked), and a rotation refused for the RAM cap is logged once per park; the head's notices are unchanged, it is restored when a resident finishes.
Also from #184: a started slot's reservation counts from the prefix it keeps, as its used count already did.

While parked, a shorter ping interval the request asked for is kept: the parked keepalive is the shorter of the configured interval and the forced two seconds.

From #184 again: a started slot is counted by the cells it holds, and trimmed to the prefix its request keeps when nothing fits.

A started slot is validated before it can be parked. The prompt checks of the STARTED block (logits without memory, a prompt past the physical batch or the context) are one helper now, run there as before and asked by the planner before a started slot can be chosen, so a request the checks would reject is never given a park notice ahead of its error: the notice opens the stream, and the client would have got 200 plus an in-stream error where the non-stream 4xx belongs.

### CODE REVIEW

**Comment by danielhanchen** [comment]
One more commit, `5a791e03c`: the keepalive this PR promises never actually arrived, and the reason is in the result queue rather than in the notice path.

`server_response::send()` calls `notify_all()` on a single condition variable for every result of every task, and `server_response::recv_with_timeout()` waited with `wait_for()`, which restarts the wait on each wakeup. A reader waiting on a task that is producing nothing is therefore woken by every token every other task produces and its timeout never elapses, so on a server with any traffic at all the timeout is not a timeout. A parked slot is the worst possible case by construction, because a slot is only ever parked while the others are busy, which is exactly when the wakeups never stop. The same applies to the ordinary `--sse-ping`, which could only ever fire on an otherwise idle server, and to the `should_stop` polling in `server_response_reader::next()`, whose own comment says it happens every `polling_interval_seconds` and did not. The fix is to compute the deadline once and `wait_until()` it, so a spurious wakeup re-checks the queue and goes back to waiting for the same instant, which is what every caller already reads the argument as meaning.

Measured with two streaming completions on the 4B at `-c 2560` with two slots, sized so they do not fit together and one is parked until the other finishes, every SSE line timestamped as it arrives. Before, on an isolated copy of `65f8f5464` with only this file reverted, the parked stream shows a park at 17.40 s and a resume at 31.27 s and nothing in between, a 13.87 s silence on a stream whose entire purpose is to say something every 2 s. After, on the same workload, the park lands at 17.39 s and the resume at 31.25 s with keepalives at 20.39, 23.39, 26.39 and 28.39 s, four across a 13.86 s park, spaced at the 2 s period plus the reader's 1 s polling granularity, and none outside the park. Both runs produced 1201 tokens on each stream.

Cost is none measurable: nothing extra happens per result, and on the branch this was found on the same four-chat control gives 139.6 tok/s before the change and 139.2 after. There is no harness test for it, because at the context sizes the test model runs at a park lasts about as long as the keepalive period itself, and a test straddling the period it is testing would be worse than none, so this is pinned by the probe above. `tools/server/tests/unit/test_preempt_notify.py` is 4 passed and `test_preempt.py` 6 passed on the new head.

**Comment by danielhanchen** [comment]
Merged #184 up to 86845c15e (the retry ladder's last resort) and added one line for this branch: the last resort parks through the same call as the planner but did not send the stream comment the planner sends, so Studio, which shows the pause from that comment, showed nothing for a slot parked that way. Found by a four-chat 35B GUI drive with the planner off: four of four finished, nine resumes, zero pauses shown. With the notice wired the same drive shows the pauses. Suites: 9 preemption and 4 notify, 13 passed; on the merged #197 binary the same drive shows the pause on 4 of 4 chats, 12 parks and 12 resumes, 0 errors.

**Comment by danielhanchen** [comment]
@codex review

File: tools/server/server-context.cpp [hunk 1]
```diff
@@ -5086,6 +5127,11 @@ std::unique_ptr<server_res_generator> server_routes::handle_completions_impl(
                     output = format_error(res_type, res_json);
                     SRV_DBG("%s", "error received during streaming, terminating stream\n");
                     return false; // terminate on error
+                } else if (const auto * notice = dynamic_cast<server_task_result_preempt_notice*>(result.get())) {
+                    // [TAG_PREEMPT] an SSE comment: invisible to clients that do not know
+                    // about preemption, a pause indicator for the ones that do
+                    parked = notice->parked;
+                    output = parked ? ": preempted\n\n" : ": resumed\n\n";
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tools/server/server-context.cpp]: Fixed in 7fb42b582: the reader keeps the set of parked prompts and runs the keepalive while any is parked; the comment names the prompt for prompts after the first, and prompt 0 keeps the bare form single-prompt clients match on.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

File: tools/server/server-context.cpp [hunk 2]
```diff
@@ -3283,6 +3317,8 @@ struct server_context_impl {
 
             metrics.n_resume++;
 
+            send_preempt_notice(*best, false);
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tools/server/server-context.cpp]: Fixed in ab40f1614: the rotation branch sends the parked notice at issue time, the same as the victim site, and the stream test pairs every notice across a rotation.

**Comment by danielhanchen** [comment]
@codex review

File: tools/server/server-context.cpp [hunk 3]
```diff
@@ -5296,10 +5362,14 @@ std::unique_ptr<server_res_generator> server_routes::handle_completions_impl(
                 // receive subsequent results
                 bool timeout = false;
                 int64_t start_time = ggml_time_ms();
-                auto result = rd.next([&timeout, &start_time, sse_ping_interval, &effective_should_stop]() {
+                // [TAG_PREEMPT] a parked slot produces nothing for as long as the pool is
+                // full, so while parked the ping runs every 2 s regardless of --sse-ping and
+                // is named, so a client can tell "waiting for cells" from "slow".
+                const int64_t ping_ms = parked ? PREEMPT_KEEPALIVE_MS : (sse_ping_interval > 0 ? (int64_t) sse_ping_interval * 1000 : -1);
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply tools/server/server-context.cpp]: Right. While parked the keepalive ran every two seconds whatever the request had asked for, lengthening the silence for a client that wanted one every second. The parked interval is now the shorter of the configured one and the forced two seconds. Fixed in 02a908bf2.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex review
